nova_plugin/host_aggregate.py

```diff
@@ -52,25 +55,29 @@ def _set_metadata(ctx, nova_client, host_aggregate, kwargs):
 
     if metadata:
         ctx.logger.debug(
-            'Adding metadata {0} to aggregate {1}'
+            'Setting metadata {0} for aggregate {1}'
             .format(metadata, host_aggregate)
         )
 
         nova_client.aggregates.set_metadata(host_aggregate, metadata)
 
 
-def _remove_hosts(ctx, nova_client, host_aggregate_id, kwargs):
-    hosts = ctx.instance.runtime_properties.get(HOSTS_PROPERTY, [])
-
+def _remove_hosts(ctx, nova_client, host_aggregate_id, hosts):
     for host in hosts:
         ctx.logger.debug(
             'Removing host {0} from aggregate {1}'
             .format(host, host_aggregate_id)
         )
-
         nova_client.aggregates.remove_host(host_aggregate_id, host)
 
-    ctx.instance.runtime_properties.pop(HOSTS_PROPERTY, None)
+    ctx.instance.runtime_properties[HOSTS_PROPERTY] = [
+        host
+        for host in ctx.instance.runtime_properties.get(HOSTS_PROPERTY, [])
+        if host not in hosts
+    ]
+
+    if len(ctx.instance.runtime_properties.get(HOSTS_PROPERTY, [])) == 0:
```

Comment:
maybe `if not ctx.instance.runtime_properties.get(HOSTS_PROPERTY):` is more readable